### DISCUSSION TITLE
NFS Loadbalance Service

### DISCUSSION THREAD

**Discussion by kyle-rf** [rook/rook#7610] [Q&A] [answered]
Hello,

I do not have much experience with NFS.  So maybe there is a constraint I am not aware of, but I am curious why Rook provisions NFS active servers as separate deployments and services.  I would like to have three nfs pods load-balanced with a single service so pods can fail and be replaced without having to manually change anything for client pods and external clients that use it.

Is there a specific reason why this can't be done and/or why Rook deploys it as separate deployments/services?

**Comment by richardelling** [comment, marked as answer]
This question has many dimensions, so it might help if you explain what you are trying to accomplish in more detail.

For the NFSServer CRD, which deals with NFS v3 and v4 services, those versions of the NFS protocol are not suited to multiple instances running against a single backing filesystem.

For the CephNFS CRD, which deals with NFS v4.1, it is possible to have multiple NFS server instances working in parallel (with a large number of constraints)

However, it seems like you are describing a problem that doesn't really exist, namely that a NFS server pod can fail and cause clients to require changes. This is not the case. NFS was designed in the days when servers were not clustered and the networks were not reliable, so it is very tolerant of disruptions. The clients, with proper configuration, will patiently retry and when the NFS pod is back up, will resolve outstanding locks and continue. The Service does not require any changes, it will find the NFS pod on whatever node it lands upon. For external consumers, the load balancer will ensure the traffic is routed to the Service properly.

Finally, NFS service is I/O intensive, but not CPU intensive. So it rarely helps to spread across servers in an effort to throw more CPUs at NFS. YMMV.

There are hundreds of ways to configure all of this, so it won't be easy to offer a more detailed answer without clear requirements and constraints.

**Reply by kyle-rf** [reply]
Thank you for the information Richard.  It gives me a little more understanding than I had before.

I am using CephNFS.  It makes sense to me that there shouldn't be any issues for clients running inside the K8s cluster.  Since they can reference it with the cluster's internal DNS.  However, with my external clients, there is a problem.  The CephNFS resource deploys a ClusterIP service with ClusterIP set to None and there is no way, AFAIK, to change it in the resource definition or after it is deployed.  So I was using the IP of the node that the nfs pod gets deployed on (For my external clients).  That's where I am running into this issue, when the pod ends up moving.  I have to change the clients when this happens.  I could set a DNS entry, but I'd still have to update the DNS record to the new node's IP anytime the pod moves.  I could create my own service for it, that is a LoadBalancer type as well, but I'd rather be able to have the CephNFS be able to set it. 

I see a Feature Request was created recently: https://github.com/rook/rook/issues/7674

I think not being able to deploy CephNFS with the service as a LoadBalancer is what was tripping me up.  But it had me questioning whether I can/should loadbalance NFS across multiple endpoints.  From your explanation it sounds like I probably shouldn't, and simply set a LoadBalancer service with a single endpoint.  That will eliminate my issue of having to reconfigure my external consumers.

**Reply by richardelling** [reply]
yes, that does make sense. The bad news is that there are a lot of different LoadBalancers out there, each with their own special configs and ways of handling IP address assignment. Thus it would be difficult to come up with a generic method of adding this into an orchestrator like rook. The good news is that creating a Service that uses a LoadBalancer is relatively straightforward. For CephNFS, you'll want to expose an IP address for each CephNFS server. There is no need to load balance across them all, because they are already running NFS v4.1 and the clients know how to failover.

**Reply by kyle-rf** [reply]
Gotcha, that makes sense.  Thanks again!
